Added listing of all the commands.

wqyeo opened the pull request:
Based on #3 

`help` just simply lists out all the commands in an embed.<br>The user can navigate through the different pages using the LeftArrow and RightArrow emoji.

*Alternatively*, the user can also filter the search by `help (filterHere)`.<br>In-code, it just uses `string.Contains` on the `Summary`, `Remarks` and `Name` attribute if applicable.

Pages are separated based on modules rather than length.

>The bot will just say a warning if the filter failed to find any respective commands.
The bot will not show the respective emojis if the commands can fit into a page.

### Code Design
**For handling Message Reactions**

`CommandHandler` makes the respective static `MessageReactionHandler` function subscribe to the respective `Client` object reaction event-handlers.

To make an object have a callback when a reaction-related event occurs to a message, let it implement `IMessageReactionHandle` and call `AddMessageReactionHandle` from `MessageReactionHandler`, with the input being the respective object.

Just remember to call `RemoveMessageReactionHandle` after you are done, with the input being the respective object.

Commit:
Merge branch 'Issue_#42'

Commit:
Merge branch 'Issue_#42'

Commit:
HelpService now has a function to return a list of help pages as embed.

Commit:
Added MessageReactionHandler

A handler that callbacks to whatever that is subscribed to it when a reaction is removed/added/cleared from a message

Commit:
Added HelpBook object;

Will handle the reaction events for help command

Commit:
Subscribed the helpbook object to MessageReactionHandler event

Commit:
Bug Fixing and small code refactor

Commit:
Added command search filter

Commit:
Added timer before disposing of the HelpBook;

Code refactor

Commit:
Forgot to un-comment some codes (for testing reasons);

Darius-Wattimena commented on Cake.Core/Discord/Modules/HelpModule/HelpModule.cs:
```diff
@@ -25,22 +25,32 @@ public HelpModule(HelpService service, CommandService commandService)
+            else
+            {
+                await newMessage.ModifyAsync(msg => { msg.Embed = helpPages[0].Build(); msg.Content = string.Empty; });
 
```
Here might be a better place to do a check if the count != 1

This way you can remove the else if you have right now

Darius-Wattimena commented on Cake.Core/Discord/Handlers/CommandHandler.cs:
```diff
@@ -52,16 +56,18 @@ private async Task PrefixCommandAsync(ShardedCommandContext context, int argPos
+                //var guild = await Database.Queries.GuildQueries.FindOrCreateGuild(context.Guild.Id);
 
-                if (context.Message.HasCharPrefix(Convert.ToChar(guild.Prefix), ref argPos))
+                if (context.Message.HasCharPrefix(Convert.ToChar(">"/*guild.Prefix*/), ref argPos))
```
Set back stuff like how it was first

Commit:
Refactored conditional statements

Kaynn-Cahya commented:
> Here might be a better place to do a check if the count != 1
> 
>This way you can remove the else if you have right now

👍

Commit:
Quick bug fix; Made sure the quotebook only responds to targeted message

Commit:
Code Refactor; Placed repeating code into function

CakeAndBanana commented:
Hello, Sorry for the late response (we were on vacation in Germany).

I was testing out the helpbook function and if you go from last page to next page, it goes to the second last page. Also can you add page numbers at the bottom as a footer.

Commit:
Commit based on Requested Changes on pull request

Footer now contains page number;
Fixed bug relating to "next page on last page brings to second-last page";
Added spacer before the command description

Kaynn-Cahya commented:
@CakeAndBanana  👍 
Hopefully I did not misinterpret your meaning of `add a spacer before the description text of the command?`

CakeAndBanana commented:
Uhm, The page number is going backwards. When I'm going forward in the pages it goes from 3 to 2 and when I'm going backwards it's going from 2 to 3. Ignore the spacer part, I can't explain it well enough 😄

Commit:
Made helpbook flip from Right-to-Left;

Based on review request

Kaynn-Cahya commented:
@CakeAndBanana 

> Uhm, The page number is going backwards. When I'm going forward in the pages it goes from 3 to 2 and when I'm going backwards it's going from 2 to 3.

Fixed it; 
P.S, I was thinking about books that flipped Left-to-Right 😅

CakeAndBanana commented:
Thanks for this PR, Everything looks fine.
I will push this to master :)
